src/abi/abi_type.ts

```diff
@@ -0,0 +1,846 @@
+/* eslint-disable no-bitwise */
+/* eslint-disable no-use-before-define */
+/* eslint-disable class-methods-use-this */
+
+/**
+    //ABI-Types: uint<N>: An N-bit unsigned integer (8 <= N <= 512 and N % 8 = 0).
+    // | byte (alias for uint8)
+    // | ufixed <N> x <M> (8 <= N <= 512, N % 8 = 0, and 0 < M <= 160)
+    // | bool
+    // | address (alias for byte[32])
+    // | <type> [<N>]
+    // | <type> []
+    // | string
+    // | (T1, ..., Tn)
+*/
+import algosdk from '../..';
+import { concatArrays } from '../utils/utils';
+
+export const MAX_LEN = 2 ** 16 - 1;
+export const ADDR_BYTE_SIZE = 32;
+export const SINGLE_BYTE_SIZE = 1;
+export const SINGLE_BOOL_SIZE = 1;
+export const LENGTH_ENCODE_BYTE_SIZE = 2;
+
+export interface Segment {
+  left: number;
+  right: number;
+}
+
+const staticArrayRegexp = /^([a-z\d[\](),]+)\[([1-9][\d]*)]$/;
+const ufixedRegexp = /^ufixed([1-9][\d]*)x([1-9][\d]*)$/;
+
+type ABIValue = boolean | number | bigint | string | Uint8Array | ABIValue[];
+
+export abstract class Type {
+  // Converts a Type object to a string
+  abstract toString(): string;
+  // Checks if two Type objects are equal in value
+  abstract equal(other: Type): boolean;
+  // Checks if the Type object (or any of its child types) have dynamic length
+  abstract IsDynamic(): boolean;
+  // Returns the size of the Type object in bytes
+  abstract ByteLen(): number;
+  // Encodes a value for the Type object using the ABI specs
+  abstract Encode(value: ABIValue): Uint8Array;
+  // Decodes a value for the Type object using the ABI specs
+  abstract Decode(byteString: Uint8Array): ABIValue;
+  // De-serializes the ABI type from a string using the ABI specs
+  static Of(str: String): Type {
+    if (str.endsWith('[]')) {
+      const arrayArgType = Type.Of(str.slice(0, str.length - 2));
+      if (arrayArgType === null) {
```

Comment:
I don't think the return type will ever be null, will it?

Reply:
Right, good catch. It should throw an error if there is nothing before it anyway, i.e. `[]`